compiler/qsc_partial_eval/src/lib.rs

```diff
@@ -2260,6 +2264,98 @@ impl<'a> PartialEvaluator<'a> {
+                            args_span,
+                        ));
+                    };
+                    input_type.push(qsc_rir::rir::Ty::Qubit);
+                    operands.push(self.map_eval_value_to_rir_operand(&Value::Qubit(*qubit)));
+                }
+            }
+            _ => {
+                return Err(Error::EvaluationFailed(
+                    "the arguments of a measurement should be Qubits".to_string(),
+                    args_span,
+                ))
+            }
+        }
+
+        match &callable_decl.output {
+            qsc_fir::ty::Ty::Prim(qsc_fir::ty::Prim::Result) => {
+                input_type.push(qsc_rir::rir::Ty::Result);
+                let result_value = Value::Result(self.resource_manager.next_result_register());
+                let result_operand = self.map_eval_value_to_rir_operand(&result_value);
+                operands.push(result_operand);
+                results_values.push(result_value);
+            }
+            qsc_fir::ty::Ty::Tuple(outputs) => {
+                for output in outputs {
+                    if matches!(output, qsc_fir::ty::Ty::Prim(qsc_fir::ty::Prim::Result)) {
+                        input_type.push(qsc_rir::rir::Ty::Result);
+                        let result_value =
+                            Value::Result(self.resource_manager.next_result_register());
+                        let result_operand = self.map_eval_value_to_rir_operand(&result_value);
+                        operands.push(result_operand);
+                        results_values.push(result_value);
+                    } else {
+                        return Err(Error::EvaluationFailed(
+                            "the arguments of a measurement should be Qubits".to_string(),
+                            args_span,
+                        ));
+                    }
+                }
+            }
+            _ => {
+                return Err(Error::EvaluationFailed(
+                    "the outputs of a measurement should be Results".to_string(),
+                    args_span,
+                ));
+            }
+        }
+
+        let measurement_callable = Callable {
+            name: callable_decl.name.name.to_string(),
+            input_type,
+            output_type: None,
+            body: None,
+            call_type: CallableType::Measurement,
+        };
+
+        // Check if the callable has already been added to the program and if not do so now.
+        let measure_callable_id = self.get_or_insert_callable(measurement_callable);
+        let instruction = Instruction::Call(measure_callable_id, operands, None);
+        let current_block = self.get_current_rir_block_mut();
+        current_block.0.push(instruction);
+
+        match results_values.len() {
+            0 => Err(Error::EvaluationFailed(
+                "a measurement should return one or more results".to_string(),
+                args_span,
+            )),
```

Comment:
```suggestion
            0 => panic!("unexpected unitary measurement"),
```

Reply:
Since this is checked at the HIR level, this should probably be a panic instead since it is an invalid state that should not happen, and if it happens it means something went wrong.